### DISCUSSION TITLE
FK for relationship property set to null on insert

### DISCUSSION THREAD

**Discussion by stevepentland** [sqlalchemy/sqlalchemy#8259] [Usage Questions] [answered]
Hi there! I know this must have been asked before but I've searched high and low and am having a hard time finding anything. It may also be I don't exactly know what to search for, but I've searched the discussions and closed issues for foreign key stuff and still don't know. I'm hoping someone can help me understand why I am seeing this behaviour. I'm sure it's pretty straightforward to someone who is more familiar with using sqlalchemy

Let's say I have the following dataclasses:
```python3
@dataclass()
class ModelBase:
    id: Optional[UUID]
    created_at: Optional[datetime]
    updated_at: Optional[datetime]

@dataclass()
class User(ModelBase):
    user_name: str
    email_address: str
    full_name: str
    message_box_id: Optional[UUID] = None
    message_box: Optional[MessageBox] = None
    login: Optional[Login] = None


@dataclass()
class Login(ModelBase):
    user_id: UUID
    hashed_password: str
    user: Optional[User] = None
```

And the tables for the latter two are created as follows:
```python3
user_table = Table(
        "users",
        r.metadata,
        Column(
            "id",
            UUID(as_uuid=True),
            primary_key=True,
            server_default=text("gen_random_uuid()"),
        ),
        Column("user_name", String(32), nullable=False, unique=True, index=True),
        Column("email_address", String(64), nullable=False, unique=True, index=True),
        Column("full_name", String(256), nullable=False),
        Column(
            "created_at",
            DateTime,
            nullable=False,
            server_default=text("timezone('utc', now())"),
        ),
        Column(
            "updated_at",
            DateTime,
            nullable=False,
            server_default=text("timezone('utc', now())"),
            server_onupdate=FetchedValue(),
        ),
        Column(
            "message_box_id",
            UUID(as_uuid=True),
            ForeignKey("message_boxes.id"),
            nullable=False,
        ),
    )

    login_table = Table(
        "logins",
        r.metadata,
        Column(
            "id",
            UUID(as_uuid=True),
            primary_key=True,
            server_default=text("gen_random_uuid()"),
        ),
        Column("user_id", UUID(as_uuid=True), ForeignKey("users.id"), nullable=False),
        Column("hashed_password", String, nullable=False),
        Column(
            "created_at",
            DateTime,
            nullable=False,
            server_default=text("timezone('utc', now())"),
        ),
        Column(
            "updated_at",
            DateTime,
            nullable=False,
            server_default=text("timezone('utc', now())"),
            server_onupdate=FetchedValue(),
        ),
    )

    r.map_imperatively(
        User,
        user_table,
        properties={
            "login": relationship("Login", back_populates="user", uselist=False),
            "message_box": relationship("MessageBox", uselist=False),
        },
    )
    r.map_imperatively(
        Login,
        login_table,
        properties={
            "user": relationship("User", back_populates="login", uselist=False)
        },
    )
```

Now, let's say I have created a `User`, and have added it to the database (via commit/flush) and have only retained the `id` value as a `UUID`. When I try to create a new `Login` type like so:

```python3
def test_add_login(user_id: UUID, session: Session):
    login = Login(
        id=None,
        created_at=None,
        updated_at=None,
        user_id=user_id,
        hashed_password=random_of_length(24)
    )
    repository = LoginRepository(session)

    # .add() is just session.add() at the moment
    repository.add(login)

    session.commit()

    assert login.id

    stored = session.get(Login, login.id)

    assert stored
    assert stored.id
    assert stored == login
```

In this context `user_id` is a proper instantiated `UUID` instance (double checked through pdb). And the `login` instance as sent to `session.add()` is like so:
```python
Login(id=None,
      created_at=None,
      updated_at=None,
      user_id=UUID('360f96ff-ace7-4a27-8f83-f6109cfffd09'),
      hashed_password='DNCRJJBQCDZWU7591X3QCM5B',
      user=None)
```

However once `commit()` is called I am met with the following error:
```
E       sqlalchemy.exc.IntegrityError: (psycopg2.errors.NotNullViolation) null value in column "user_id" of relation "logins" violates not-null constraint
E       DETAIL:  Failing row contains (6facef17-2bef-43db-bc03-d8d789961b22, null, DNCRJJBQCDZWU7591X3QCM5B, 2022-07-16 02:49:29.399237, 2022-07-16 02:49:29.399237).
E       
E       [SQL: INSERT INTO logins (user_id, hashed_password) VALUES (%(user_id)s, %(hashed_password)s) RETURNING logins.id]
E       [parameters: {'user_id': None, 'hashed_password': 'DNCRJJBQCDZWU7591X3QCM5B'}]
```

It seems like this only happens if I only provide the FK to the entity, but this is also something I would expect to be able to do, adding a FK for something that already exists in the DB without having to provide the actual instance of the entity being referenced. 

Thanks for reading all of this! Hopefully someone can help me figure out what I'm doing wrong.

**Comment by zzzeek** [comment, marked as answer]
hi -

because you are using dataclasses, your Login class is likely setting Login(user=None) as part of the constructor, which then supersedes the user_id value you're passing in.      if you are sending user_id to the constructor and not user, for dataclass support you likely need to remove "user" from your init method using field(init=False).

**Reply by zzzeek** [reply]
you can also `del login.user` to get rid of it, or use an event like [init](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=instanceevents+init#sqlalchemy.orm.InstanceEvents.init) to do the same autoamtically if "user_id" is present and "user" is None.

dataclasses presents a simplistic and "acts mostly atomic" kind of data structure, whereas manipulating user_id directly is dealing with persistence details, so the two different patterns can't be reconciled without extra steps.

**Reply by stevepentland** [reply]
Oh my gosh, that did it! Thank you so much @zzzeek I had totally overlooked that and now that I look back into my snippets in my original question I can see that it indeed is set to `None`. I feel that using `field(init=False)` will be the best bet for me here as they are not frozen and it can be dealt with later while still giving me the ability to reference existing DB contents without having to fetch them first. 

~~I guess one think I need to watch out for is if sqlalchemy tries to use the `__init__` on the respective class to set the referenced instance at query time. Or does it instead use dot properties to set the values on the returned values?~~
Edit: nvm, some quick testing showed this is not a concern

**Reply by zzzeek** [reply]
great.  2.0 has a lot of dataclasses support and I've been running up against this thing as well, not sure what the best way to deal with it yet.  it would be nice if dataclasses had the option to not assign a variable that isn't passed, but is otherwise accepted.

**Reply by stevepentland** [reply]
Yeah it's tricky. In my case I had to provide both the `init=False` but also `default=None` to `field` to make the data classes work smoothly. Otherwise the unassigned field raises an error when you try to deal with the data class after instantiation. But this also allows sqlalchemy to not be confused by a `None` during normal operation

**Comment by tbrugere-nrjx** [comment]
For anyone ending up on this conversation, I have a solution that doesn’t involve setting relationship to `init=False`, so you can still specify the relationship when instantiating 

It involves creating a special `UNSET` value, and setting the default to that instead

```python

class UNSET():
   pass

class Base(AsyncAttrs, MappedAsDataclass, DeclarativeBase, kw_only=True, repr=True):
    def __setattr__(self, name, value):
        if isinstance(value, UNSET):
            return
        super().__setattr__(name, value)

class MyTable(Base, kw_only=True):
    __tablename__ = 'my_table'
    __table_args__ = (
        PrimaryKeyConstraint('id', name='my_table_pkey'),
        ForeignKeyConstraint(['foreign_key'], ['other.id'], name='foreign_key_constraint'),
    )

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, server_default=text('gen_random_uuid()'), default=None)
    foreign_key : Mapped[uuid.UUID] = mapped_column(Uuid,)

    other: Mapped[Other] = relationship('Other', foreign_key=[foreign_key], default=UNSET())

```

Now you can still create the object populating the foreign key through the relationship 

```python
obj = MyTable(foreign_key=None, other = Other())
```

**Reply by CaselIT** [reply]
it will be fixed by 2.1 also: https://docs.sqlalchemy.org/en/21/changelog/migration_21.html#orm-mapped-dataclasses-no-longer-populate-implicit-default-collection-based-default-factory-in-dict

**Reply by tbrugere-nrjx** [reply]
oh that’s nice
